src/algorithms/configuration/maps/occupancy_grid_map.py

```diff
@@ -25,51 +32,138 @@ def __init__(self,
                  goal: Optional[Goal] = None,
                  weight_bounds: Optional[Tuple[Real, Real]] = None,
                  traversable_threshold: Optional[Real] = None,
-                 services: Services = None) -> None:
-        super().__init__(None, services)
+                 unmapped_value: Optional[Real] = None,
+                 services: Services = None,
+                 mutable: bool = True) -> None:
+        super().__init__(services=services, mutable=mutable)
         self.agent = agent
         self.goal = goal
         self.weight_grid = None
-        self.traversable_threshold = self.DEFAULT_TRAVERSABLE_THRESHOLD
+        self.traversable_threshold = None
         if weight_grid is not None:
-            self.set_grid(weight_grid, weight_bounds)
+            self.set_grid(weight_grid, weight_bounds, traversable_threshold, unmapped_value)
 
     @property
     def weight_bounds(self):
         return (0, 1)
 
-    def set_grid(self, weight_grid: List[Any], weight_bounds: Optional[Tuple[Real, Real]] = None, traversable_threshold: Optional[Real] = None) -> None:
-        self.size = Size(*array_shape(weight_grid))
-        self.grid = np.empty(self.size, dtype=np.uint8)
-        self.weight_grid = np.empty(self.size, dtype=np.float32)
+    @staticmethod
+    def __normalise(weight_bounds, x):
+        y = (weight_bounds[1] - weight_bounds[0])
+        return max(weight_bounds[0], min(weight_bounds[1], ((x - weight_bounds[0]) / y) if y else weight_bounds[0]))
+
+    def set_grid(self, weight_grid: List[Any], weight_bounds: Optional[Tuple[Real, Real]] = None, traversable_threshold: Optional[Real] = None, unmapped_value: Optional[Real] = None) -> None:
+        if not isinstance(weight_grid, np.ndarray):
+            self.set_grid(np.array(weight_grid), weight_bounds, traversable_threshold, unmapped_value)
+            return
+
+        (self.__new_grid if self.size is None else self.__update_grid)(weight_grid, weight_bounds, traversable_threshold, unmapped_value)
```

Comment:
This is clever, but a little unreadable, can we either have an if statement and two normal function calls, or declare a function pointer and then call it?

Reply:
Fair enough, changed.